src/bootstrap/flock.rs

```diff
@@ -0,0 +1,206 @@
+//! Handles creating a file lock to avoid race conditions[1].
```

Comment:
This is now the third flock implementation in-tree, there's also one for rustdoc ... maybe we should separate this out into a tiny crate or something? It would be hard to get cargo to use it since it's a submodule, but we could at least share it with rustdoc.

cc @rust-lang/cargo, do you have opinions on this? Would you be willing to separate the flock implementation into a different crate, and maybe put it on rust-lang/flock? I don't think there's any need to publish to crates.io but it would be nice to share at least between different tools.
cc @kinnison @rbtcollins, I'm sure rustup has its own implementation of flock too.

Reply:
https://github.com/rust-lang/rustup/issues/988#issuecomment-582314389 and the top level summary are probably relevant.

tl;dr: don't want the cargo implementation, it is buggy; but also haven't gotten around to fixing the issue in rustup, which is complected with https://github.com/rust-lang/rustup/issues/2441 making this a little gnarly to think through.

Reply:
Ok. x.py itself probably doesn't need an airtight flock implementation, I wouldn't expect people to be using NFS for builds. I totally get that rustup does have to deal with those scenarios though.

Does rustup's flock implementation have any downsides compared to cargo's? Could we start from that as the shared codebase instead?

Reply:
There's also fd-lock, which worked relatively well for me in the past (admittedly it was a small project). https://crates.io/crates/fd-lock/

That said, perhaps you need more control than it gives.